src/wiktextract/extractor/simple/pronunciation.py

```diff
@@ -0,0 +1,284 @@
+import re
+from typing import Optional, Union
+
+from wikitextprocessor import LevelNode, TemplateArgs, WikiNode
+from wiktextract import WiktextractContext
+from wiktextract.page import clean_node
+from wiktextract.wxr_logging import logger
+
+from .models import Sound, WordEntry
+from .parse_utils import PANEL_TEMPLATES
+from .simple_tags import simple_tag_map
+from .text_utils import POS_STARTS_RE
+
+DEPTH_RE = re.compile(r"([*:;]+)\s*(.*)")
+
+REMOVE_HYPHENATION_RE = re.compile(r"(?i)\s*hyphenation\s*,?:?\s*(.+)")
+
+def process_pron(
+    wxr: WiktextractContext,
+    pr_nodes: list[Union[str, WikiNode]],
+    target_data: WordEntry,
+) -> None:
+    """Process a Pronunciation section WikiNode, extracting Sound data entries
+    which are inserted into target_data.sounds. target_data is a WordEntry, so
+    can be base_data (used to complete other entries) or an individual POS
+    entry."""
+
+    # XXX: figure out a way to collect category here with clean_node so that
+    # it can be properly assigned to the right POS WordEntry; currently,
+    # clean_node insert the category stuff straight into whatever data object
+    # is its target, which would make target_data (if used for this) basically
+    # have every category from every pron-section level. We already use a hack
+    # to later assign sound data to the correct POS with the `pos` field in
+    # SoundEntry... Currently ignoring categories for sound.
+
+    # print("====")
+    # print_tree(pr_node)
+
+    # We save data in parse_pronunciation_template_fn into this local list,
+    # so the template_fn has to be defined inside this larger function so
+    # that it has easy access to sound_templates. Can't be defined outside
+    # this function either, because we need access to `wxr` from here, and
+    # the template_fn signature is already set in wikitextprocessor.
+    sound_templates: list[Sound] = []
+
+    # Template handling functions are function objects that are used to process
+    # a parsed template node (a WikiNode object parsed from `{{template|arg}}`
+    # that hasn't been expanded yet into text). The function objects are passed
+    # into certain functions like like clean_node. When the expander comes
+    # across a template, before expanding the template it calls whatever was
+    # passed to template_fn=. The handler function can return a string, which is
+    # inserted into the returned text, or None in case nothing special will be
+    # done with the template and it will use the normal expanded value.
+    # post_template_fn= is the same, except it happens after the template has
+    # already been expanded and receives that text as a parameter. If you want
+    # to replace templates based on their names or arguments before expansion,
+    # use template_fn=, if you want access to the expanded text without doing
+    # it manually, use post_template_fn=
+    def parse_pronunciation_template_fn(
```

Comment:
I'm a bit disagree with this `clean_node()` template process functions practice. Normally we need the parsed expanded template node structure and its parameters.

```python
expanded_node = wxr.wtp.parse(
    wxr.wtp.node_to_wikitext(template_node), expand_all=True
)
```

and also we could use [`TemplateNode.template_parameters()`](https://github.com/tatuylonen/wikitextprocessor/blob/272742c89ec9a82208022d5f6063a1736c90b7fb/src/wikitextprocessor/parser.py#L569) method if use template args is easier.

I'd recommend create function for each template, create extracted `Sound` data directly in these functions and avoid processing plain text and long nested functions.

Reply:
I think it's because I am so influenced by the messiness of English Wiktionary source wikitext. The main English Wiktionary extractor has to be a complex mess because the source is a complex mess, and you are correct that other extractors can get away with much simpler code. I'll take a look at how it's done on Simple Wiktionary and whether you can trust the source to be "stable".

Reply:
I totally understand and agree with you the en code situation...

I think we're probably fine with processing expanded wikietxt and HTML nodes, even if the template changed dramatically it's still manageable, at least couldn't be worse then processing plain text.